File: backend/app/models/fitness.py
from sqlalchemy import Column, Integer, String, Text, Date, Float, Boolean, ForeignKey

from app.database import Base


class Workout(Base):
    __tablename__ = "workouts"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"))
    date = Column(Date, nullable=False)
    type = Column(String(100), nullable=False)
    duration_minutes = Column(Integer, nullable=False)
    calories = Column(Integer, nullable=True)
    notes = Column(Text, nullable=True)


class FitnessGoal(Base):
    __tablename__ = "fitness_goals"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"))
    name = Column(String(200), nullable=False)
    target = Column(Float, nullable=False)
    current = Column(Float, default=0.0)
    unit = Column(String(50), nullable=True)


# ── Fitness Hub models (99-phase plan, Phases 9-14) ──

class Exercise(Base):
    """A single exercise with its set/reps/weight prescription (Phase 9)."""
    __tablename__ = "exercises"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String(200), nullable=False)
    muscle_group_id = Column(Integer, ForeignKey("muscle_groups.id"))
    sets = Column(Integer, default=3)
    reps = Column(Integer, default=10)
    weight = Column(Float, default=0.0)
    user_id = Column(Integer, ForeignKey("users.id"))


class MuscleGroup(Base):
    """Anatomical muscle group with Upper/Lower body-part tag (Phase 10)."""
    __tablename__ = "muscle_groups"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String(100), nullable=False)
    body_part = Column(String(20), default="Upper")  # Upper / Lower
    image_3d_url = Column(String(500), nullable=True)
    sort_order = Column(Integer, default=0)
    user_id = Column(Integer, ForeignKey("users.id"))


class WorkoutSplit(Base):
    """A day's PUSH/PULL/LEG split with its exercise list (Phase 11)."""
    __tablename__ = "workout_splits"

    id = Column(Integer, primary_key=True, index=True)
    day_of_week = Column(Integer, default=0)  # 0=Mon ... 6=Sun
    split_name = Column(String(20), default="PUSH")  # PUSH / PULL / LEG / REST
    exercise_list = Column(Text, nullable=True)  # JSON array of exercise names
    week_number = Column(Integer, default=1)
    user_id = Column(Integer, ForeignKey("users.id"))


class Expense(Base):
    """Fitness expense (Phase 12)."""
    __tablename__ = "expenses"

    id = Column(Integer, primary_key=True, index=True)
    title = Column(String(200), nullable=False)
    cost = Column(Float, default=0.0)
    date = Column(Date, nullable=False)
    category = Column(String(50), default="Supplement")  # Supplement / Equipment / Gym
    user_id = Column(Integer, ForeignKey("users.id"))


class PersonalRecord(Base):
    """Lift PR: current vs target weight (Phase 13)."""
    __tablename__ = "personal_records"

    id = Column(Integer, primary_key=True, index=True)
    exercise_name = Column(String(200), nullable=False)
    current_weight = Column(Float, default=0.0)
    target_weight = Column(Float, default=0.0)
    unit = Column(String(20), default="kg")
    user_id = Column(Integer, ForeignKey("users.id"))


class DietPlan(Base):
    """Diet phase: Diet / Bulking / Cutting / Maintenance (Phase 14)."""
    __tablename__ = "diet_plans"

    id = Column(Integer, primary_key=True, index=True)
    title = Column(String(100), nullable=False)
    is_active = Column(Boolean, default=False)
    sort_order = Column(Integer, default=0)
    user_id = Column(Integer, ForeignKey("users.id"))

File: frontend/src/services/api/types/fitness.ts
// Fitness hub types.
// Extracted from services/api.ts (F8 split).

export interface Workout {
  id: number; user_id: number; date: string;
  type: string; duration_minutes: number;
  calories: number | null; notes: string | null;
}

export interface FitnessGoal {
  id: number; user_id: number; name: string;
  target: number; current: number; unit: string | null;
}

// ----- Fitness Hub types (99-phase plan, Phases 43-48) -----
export interface Exercise {
  id: number; name: string; muscle_group_id: number | null;
  sets: number; reps: number; weight: number; user_id: number;
}

export interface MuscleGroup {
  id: number; name: string; body_part: string;
  image_3d_url: string | null; sort_order: number; user_id: number;
}

export interface MuscleGroupOverview extends MuscleGroup {
  exercise_count: number;
}

export interface WorkoutSplit {
  id: number; day_of_week: number; split_name: string;
  exercise_list: string | null; week_number: number; user_id: number;
}

export interface SplitDay {
  day: string; day_of_week: number;
  split_name: string; exercises: string[];
}

/** Per-day Weekly-Split actuals (audit defect #84). */
export interface WeekActualDay {
  day: string;
  day_of_week: number;
  date: string;
  workouts: { id: number; type: string; duration_minutes: number; calories: number | null }[];
  total_minutes: number;
  logged: boolean;
  vault_mentions: boolean;
}

export interface WeekActualsResponse {
  week_start: string;
  days: WeekActualDay[];
}

export interface Expense {
  id: number; title: string; cost: number;
  date: string; category: string; user_id: number;
}

export interface ExpenseSummary {
  total: number;
  by_category: Record<string, number>;
}

export interface PersonalRecord {
  id: number; exercise_name: string;
  current_weight: number; target_weight: number;
  unit: string; user_id: number;
  // computed by pr_summary helper
  percent?: number;
}

export interface DietPlan {
  id: number; title: string; is_active: boolean;
  sort_order: number; user_id: number;
}

export interface WeightGoal {
  initial: number | null; current: number | null;
  target: number | null; percent: number;
}

export interface Membership {
  membership_status: string;
  next_payment_date: string | null;
  days_to_payment: number | null;
}

export interface SpecHabitHeatmap {
  id: number; name: string; goal: string | null;
  days_completed: number; percent: number;
  heatmap_7x7: { date: string; completed: boolean; count: number }[];
}

export interface FitnessHubSummary {
  weight_goal: WeightGoal | null;
  pr_tracker: PersonalRecord[];
  membership: Membership | null;
  diet_plans: DietPlan[];
  expenses_summary: ExpenseSummary;
  weekly_split: { 1: SplitDay[]; 2: SplitDay[] };
  muscle_groups: MuscleGroupOverview[];
  spec_habits: SpecHabitHeatmap[];
}
